### PR TITLE
Don't require workspace for reading from text parameter.

### PR DESCRIPTION
**Pull request by steveprentice**
* Created abstract base classes for read steps to remove duplication.
* Updated tests to not allocate node while reading text.

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStepExecution.java [hunk 1]
```diff
@@ -38,28 +38,26 @@
...
 import java.util.Properties;
-import java.util.TreeMap;
 
 /**
  * Execution of {@link ReadPropertiesStep}.
  *
  * @author Robert Sandell &lt;rsandell@cloudbees.com&gt;.
  */
-public class ReadPropertiesStepExecution extends AbstractSynchronousNonBlockingStepExecution<Map<String, Object>> {
+public class ReadPropertiesStepExecution extends AbstractFileOrTextStepExecution<Map<Object, Object>> {
```
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStepExecution.java]: why the change of key type?

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadYamlStep.java [hunk 2]
```diff
@@ -165,6 +95,8 @@ public Step newInstance(Map<String, Object> arguments) throws Exception {
 		 */
 		@Override
 		protected Object run() throws Exception {
+			super.run();
```
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadYamlStep.java]: Another, perhaps better approach is to have another `abstract` method in `AbstractFileOrTextStepExecution` that is called from it's `run` method and then override that abstract method here. That way you don't forget to call `super.run();` in every sub class.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/AbstractFileOrTextStepDescriptorImpl.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    public AbstractFileOrTextStepDescriptorImpl(Class<? extends AbstractFileOrTextStep> stepType, Class<? extends AbstractFileOrTextStepExecution> executionType) {
+        super(executionType);
+        this.stepType = stepType;
+    }
+
+    @Override
+    public AbstractFileOrTextStep newInstance(Map<String, Object> arguments) throws Exception {
+        AbstractFileOrTextStep step = DescribableHelper.instantiate(stepType, arguments);
+        if (isBlank(step.getFile()) && isBlank(step.getText())) {
+            throw new IllegalArgumentException("At least one of file or text needs to be provided.");
```
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/AbstractFileOrTextStepDescriptorImpl.java]: Even though it has no real function, having all messages listed in `Messages.properties` would be good.

**Comment by steveprentice** [comment]
I reviewed the failed validation build as well, and it doesn't look related. The windows branch passed and the linux branch timed out. Perhaps it can be kicked off again?

**Comment by rsandell** [comment]
Just trying to tickle the PR builder

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStep.java [hunk 4]
```diff
@@ -25,74 +25,31 @@
...
  * Reads java properties formatted files and texts into a map.
  *
  * @author Robert Sandell &lt;rsandell@cloudbees.com&gt;.
  */
-public class ReadPropertiesStep extends AbstractStepImpl {
-    private String file;
-    private String text;
-    private Map<String, Object> defaults;
+public class ReadPropertiesStep extends AbstractFileOrTextStep {
+    private Map defaults;
```
[the remark this answers is not in the document]
Comment (rsandell) [reply src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStep.java]: OK, have you filed a bug report about that?

[the remark this answers is not in the document]
Comment (steveprentice) [reply src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStep.java]: @rsandell, I looked into this a bit more and it looks like it has been fixed upstream a long while ago, but the dependencies declared in this plugin use very old versions of workflow-step-api. As of version 2.0 of workflow-step-api, `newInstance()` now uses the structs plugin's `DescribableModel` class instead, which appears to have a fix for this issue since early 2016.

Is there a reason the pom.xml of this plugin still points to old versions of the workflow-step-api? If not, I can open a PR to bump them.

(Also, thanks for merging and releasing!)

[the remark this answers is not in the document]
Comment (rsandell) [reply src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStep.java]: Pure laziness and lack of time. There is more work to do than just bump the dependencies;

* The pom needs to be adapted to plugin parent 2.x
* then bump the dependencies at the same time, inclusing core to something like 2.7.3 at least
* All the `@Inject` and `@StepContextParameters` et.al. needs to be changed and adapted to the new way of doing things.

Most of those things are not strictly needed, but I prefer this plugin to be as "pure" as possible and that desire made it just a bit too overwhelming for me to find time to do it.
If you're up for it you are welcome to.
